codex/stores/maintenance.nim

```diff
@@ -10,92 +10,318 @@
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+
+      md.some
+  )
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  leavesCount: Natural,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(treeCid, leavesCount, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  cid: Cid,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(cid, 0, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc ensureExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  minExpiry: SecondsSince1970): Future[?!void] {.async.} =
+  ## Sets the dataset expiry to a max of two values: current expiry and `minExpiry`,
+  ## if a dataset for given `treeCid` is not currently tracked a CatchableError is thrown
+  ##
+
+  trace "Updating a dataset expiry", treeCid, minExpiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await modify[DatasetMetadata](self.metaDs, key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        let datasetMd = DatasetMetadata(
+          expiry: max(currDatasetMd.expiry, minExpiry),
+          leavesCount: currDatasetMd.leavesCount,
+          manifestsCids: currDatasetMd.manifestsCids,
+          checkpoint: currDatasetMd.checkpoint
+        )
+        return datasetMd.some
+      else:
+        raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " not found")
+  )
+
+
+proc recordCheckpoint*(self: DatasetMaintainer, treeCid: Cid, datasetMd: DatasetMetadata): Future[?!void] {.async.} =
+  # Saves progress or deletes dataset metadata if progress > leavesCount
+  #
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
 
-proc runBlockCheck(self: BlockMaintainer): Future[void] {.async.} =
-  let expirations = await self.repoStore.getBlockExpirations(
-    maxNumber = self.numberOfBlocksPerInterval,
-    offset = self.offset
+  await self.metaDs.modify(key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
+          raise newException(CatchableError, "Change in expiry detected, interrupting maintenance for dataset with treeCid " & $treeCid)
+
+        if currDatasetMd.checkpoint.progress > datasetMd.checkpoint.progress:
+          raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)
+
+        if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
+          DatasetMetadata.none
+        else:
+          datasetMd.some
+      else:
+        raise newException(CatchableError, "Metadata for dataset with treeCid " & $treeCid & " not found")
```

Comment:
This is more readable imo:

```suggestion
  proc modifyData(maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
    without currDatasetMd =? maybeCurrDatasetMd:
      raise newException(CatchableError, "Metadata for dataset with treeCid " & $treeCid & " not found")
      
    if currDatasetMd.expiry != datasetMd.expiry or currDatasetMd.manifestsCids != datasetMd.manifestsCids:
      raise newException(CatchableError, "Change in expiry detected, interrupting maintenance for dataset with treeCid " & $treeCid)

    if currDatasetMd.checkpoint.progress > datasetMd.checkpoint.progress:
      raise newException(CatchableError, "Progress should be increasing only, treeCid " & $treeCid)

    if currDatasetMd.leavesCount <= datasetMd.checkpoint.progress:
      DatasetMetadata.none
    else:
      datasetMd.some
      

  await self.metaDs.modify(key, modifyData)
```
However, I also have a few comments:
1. Returning DatasetMetadata.none seems like an odd way to indicate that the
   metadata should be deleted. Maybe it might be better to handle the delete
   logic later on when cleaning up?
2. Why are we raising an exception for metadata not found here? It seems like it
   would be better placed for `nim-datastore` to handle that, and this
   predicate/callback would not be called because a failure would have been
   returned further up the call stack.
3. This callback is `try/except`ed up the callstack and turned into a Result
   which eventually becomes the return value of `modify`. If we were to change
   the signature of this callback to return a Result instead of raising
   exceptions, then we know that returning `failure(err)` in the callback will
   become the returned value of `modify` and hence would be a lot easier to swallow
   as a reader.
4. Since these Results are ultimately bubbled up to `superviseDatasetDeletion`,
   it's a good idea to type them properly so that they can be inspected and
   different failures can have differnt outcomes. For example, the "change in
   expiry" that is being checked sounds like it would be a nasty bug of unknown
   origin, so you may want to add a metric to it so you can monitor the
   occurences better.

```

Reply:
1. Keep in mind that it has to be done in a concurrent safe way, otherwise some anomalies will occur in the datastore, like deleting still used data or not deleting expired data. `modify` is the only way to perform concurrent safe updates to the records and it requires to return `none` if delete is an intended result of the operation.
2. An exception is raised because it would be an error situation when we're trying to record a checkpoint when there's no `DatasetMetadata` that's related to the given `treeCid`. If you would like to see API for `modify` changed, please create an issue or maybe a PR in [nim-datastore](https://github.com/codex-storage/nim-datastore) that would explain in detail how such API would look like.
3. This PR uses existing API in `nim-datastore`. If you would like this API to be changed please raise such issue with detailed explanation in appropriate repo.
4. No, change in `expiry` would not mean a nasty bug. It would mean that there was an update to the dataset metadata, that could be a result of for example reuploading a dataset just after it expired and maintenance started but hasn't finished yet.

Reply:
I believe @dryajov's comment address points 1-3. Regarding point 4, I think you
missed the point. The comment was meant to provide reasoning about why typing
exceptions is important. But as with the change that Dmitriy suggested, a Result
would be returned and exceptions would not need to be raised here.

Reply:
Once this API proposed by @dryajov will be available in `nim-datastore` I will use it. Even though I think it's pretty much the same as the current one with additional unnecessary complications to it.